### DISCUSSION TITLE
Support private plugin repositories

### DISCUSSION THREAD

**Discussion by gewenyu99** [trunk-io/docs#276] [Feature Requests]
Originally posted on Canny by Tyler Jang:

Support private repositories through a `GITHUB_TOKEN`, `gh` client, `git` credential manager, or built-in Trunk auth flow.

Private GitHub plugin repositories are not currently supported: https://docs.trunk.io/check/advanced-setup/plugins#plugins-scope

**Comment by sheldonhull** [comment]
I got notified of this option from @TylerJang27 which I appreciate being in beta.
I'll do my best to give it a try and report back progress 🙇🏻

**Reply by TylerJang27** [reply]
Exclusively `1.22.3-beta.8` includes limited support for private plugins at this time. Specifically, support for [git-credential-manager](https://github.com/git-ecosystem/git-credential-manager/tree/release). Please let me know if you run into any issues or if you have any requests, particularly for CI compatibility as well

**Reply by TylerJang27** [reply]
@sheldonhull Have you had a chance to try this out yet?

**Reply by sheldonhull** [reply]
@TylerJang27  is this version still the one I should be referencing?
I ran an internal test using normal updated version and it prompted in browser for SSO so worked fine for that. I'm about to test with unattended to see if works equally as well. 

Also should I be changing the version of the trunk CLI in the calling repo or in the trunk-plugins/trunk-configs repos and where? Just making sure as there's a lot of abstraction so making sure I go to the right place.

Right now I'm running:

	version: 0.1
	cli:
	  version: 1.22.5


Note that the docs talk about a shared config being checked out, but when reviewing the trunk-plugins/trunk-configs repos you have, there is not a pattern I see this in. Only a `.trunk` folder, which I thought was per repo, and not shared, but maybe plugins/configs being added changes the behavior to be common? 

![CleanShot 2024-09-23 at 14 13 56](https://github.com/user-attachments/assets/3df59a04-9bcd-45ea-bdf9-f5e984a6987d)


- For shared configs: the pattern seems to be designated "exported configs".
- For shared plugins: seems the `.trunk` file points to inheriting with enabled option in the repo level `.trunk/trunk.yaml`.... 
	
		enabled:
		    # enabled actions inherited from github.com/trunk-io/configs plugin
		    - linter-test-helper
		    - npm-check-pre-push
		    - remove-release-snapshots
		    - repo-tests
		    - tool-test-helper

**Reply by TylerJang27** [reply]
The CLI should be changed in the calling repo. The `trunk-plugins` repo should not matter in this case, persay.

If you're using a private GH repository, you'll need to use `1.22.3-beta.8`. If you're using a different git hosting provider, you _should_ be able to use latest

**Reply by sheldonhull** [reply]
Ok, one thing I found that added more detail on this: [configs inheritance](https://github.com/trunk-io/configs/blob/main/contributing.md#config-merging-overview)... this might be worth being in the main docs.

**Reply by sheldonhull** [reply]
Ok, so far success locally when I manually update the plugin version. 

However, upgrade doesn't work with the remote repo, detecting the new tag version. I have to manually update.

**Reply by TylerJang27** [reply]
Ah, good callout, we'll move some of those docs over

For upgrading we query GH APIs, which means if you're private or on non-GH we don't support that currently. Were you using the beta version or 1.22.5?

**Reply by sheldonhull** [reply]
I'm using private azure devops repo.
If you are using the upstream api to query tags then that would explain it if it's not being cloned or done in a generic git way.

I did succeed after a few hours of work with azure devops but no ability out of the box to tell a team downstream `trunk upgrade` to pull in the changes. I'd have to give them a script or push those out myself since the upgrade won't honor that. 

Thanks for the help today 👍🏻

**Reply by sheldonhull** [reply]
@TylerJang27 I guess I could probably create an azure cli call to fetch upstream latest tag and essentially reproduce the effect myself using azure cli call+ yq as tools. 

Refresh me though if I've got folks on windows and Mac will an action script run in an emulated shell or whatever the user is running trunk in? Therefore I'd have to have it fall a cross platform PowerShell script or use another language runtime like Go and include a main.go to run as a custom action in my plugins?

**Reply by TylerJang27** [reply]
To clarify, you're trying to write a custom Trunk Action that mimics the behavior of fetching latest and updating the plugin ref? Trunk Actions run in a managed `bash` by default, but you can override that with `actions.definitions[].shell`

**Reply by sheldonhull** [reply]
Yes. So is that managed bash cross platform. I know of the go package that has a shell that's cross platform but limited tooling in it like mkdir and awk aren't accessible. 

Any details on docs on this managed bash so I can write some scripts that work on windows and Linux/mac?

**Reply by TylerJang27** [reply]
Sorry, I misspoke slightly. On Linux/Mac we'll inherit from your `PATH` and use the `bash` we find there (and whatever commands you call). On Windows we install from [git-bash](https://github.com/git-for-windows/git/releases) and also inherit your PATH.

You can get some additional debug info about what we run using `trunk actions history <action-id>` and looking at the YAML files. You can also set `actions.output_level: verbose` for some info as things execute

**Reply by sheldonhull** [reply]
@TylerJang27  so I setup an internal (org) repo, and confirmed it had a release with a new version.
Then running trunk upgrade it checked for updates, but did not increment the version to the new patch version. 

This is github now since the toolchain is limited to supporting this for automatic updates. 

I noticed the behavior also ended up replacing the beta flag version you mentioned, so I reran with `trunk upgrade plugins` and same behavior, no version update found.

**Reply by sheldonhull** [reply]
correction... wow it took a while to respond, probably something on my network. I did upgrade to next patch.

	Querying ReleaseVersionService failed with message Unable to find latest release version [DAEMON]
	[02-24 16:05:21] [plugin.cc:296] Got HTTP 404 curling plugin https://github.com/{ORG}/trunk-configs/archive/v0.3.19.zip, using git fallback [DAEMON]

I also noted this, which not clear on why this would come in debug logs, since I don't have it executing 1password for example, though I have the cli installed in my system.

	[02-24 16:05:26] [download_tool.hh:32] Failed to create download for tool  with error '1password-cli' has no matching download for current os/cpu @ version:2.19.0 [DAEMON]
		[02-24 16:05:26] [download_tool.hh:32] Failed to create download for tool  with error 'dotnet' has no matching download for current os/cpu @ version:7.0.400 [DAEMON]
		[02-24 16:05:26] [trigger_set.hh:53] Failed to add regex: invalid perl operator: (?! [DAEMON]

**Reply by TylerJang27** [reply]
@sheldonhull This has now been released in `1.22.11`!

**Comment by sheldonhull** [comment]
The docs saw this is still not supported, the link is dead, and I'm running into some issues right now that might be entirely just me. Can I get a confirmation that this is still supported?
